lib/streamlit/state/session_state.py

```diff
@@ -114,21 +263,73 @@ def __getattr__(self, key: str) -> Any:
     def __setattr__(self, key: str, value: Any) -> None:
         # Setting the _old_state and _new_state attributes must be done using
         # the base method to avoid recursion.
-        if key in ["_new_state", "_old_state"]:
+        if key in ["_new_session_state", "_new_widget_state", "_old_state"]:
             super().__setattr__(key, value)
         else:
             self[key] = value
 
     def __delattr__(self, key: str) -> None:
+        if key in ["_new_session_state", "_new_widget_state", "_old_state"]:
+            raise AttributeError(key)
+
         try:
             del self[key]
         except KeyError:
             raise AttributeError(key)
 
+    def _set_from_proto(self, widget_states: WidgetStatesProto):
+        for state in widget_states.widgets:
+            self._new_widget_state.set_from_proto(state)
+
+    def _call_callbacks(self):
+        changed_widget_ids = [
+            wid for wid in self._new_widget_state if self._widget_changed(wid)
+        ]
+        self._compact_state()
+        for wid in changed_widget_ids:
+            self._new_widget_state.call_callback(wid)
+
+    def _widget_changed(self, widget_id: str) -> bool:
+        new_value = self._new_widget_state.get(widget_id, None)
+        old_value = self._old_state.get(widget_id, None)
+        changed: bool = new_value != old_value
+        return changed
+
+    def _reset_triggers(self) -> None:
+        """Sets all trigger values in our state dictionary to False."""
+        for state_id in self._old_state:
+            metadata = self._new_widget_state.widget_metadata.get(state_id, None)
+            if metadata is not None:
+                if metadata.value_type == "trigger_value":
+                    self._old_state[state_id] = False
+
+    def _cull_nonexistent(self, widget_ids: Set[str]):
+        self._new_widget_state.cull_nonexistent(widget_ids)
+
+    def _set_metadata(self, widget_metadata: WidgetMetadata) -> None:
+        widget_id = widget_metadata.id
+        self._new_widget_state.widget_metadata[widget_id] = widget_metadata
+
+    def _get_value_for_registration(self, widget_id: str) -> Any:
+        try:
+            value = self[widget_id]
+            return value
+        except KeyError:
+            metadata = self._new_widget_state.widget_metadata.get(widget_id, None)
+            if metadata is None:
```

Comment:
minor, but I don't believe this case is possible since this method is only called after widget metadata is registered